core/domain/skill_services.py

```diff
@@ -230,6 +235,251 @@ def get_all_skill_summaries():
     return skill_summaries
 
 
+def _get_skill_summaries_in_batches(
+        num_skills_to_fetch, urlsafe_start_cursor, sort_by):
+    """Returns the summaries of skills present in the datastore.
+
+    Args:
+        num_skills_to_fetch: int. Number of skills to fetch.
+        urlsafe_start_cursor: str or None. The cursor to the next page.
+        sort_by: str. A string indicating how to sort the result.
+
+    Returns:
+        3-tuple(skill_summaries, new_urlsafe_start_cursor, more). where:
+            skill_summaries: list(SkillSummary). The list of skill summaries.
+                The number of returned skill summaries might include more than
+                the requested number. Hence, the cursor returned will represent
+                the point to which those results were fetched (and not the
+                "num_skills_to_fetch" point).
+            urlsafe_start_cursor: str or None. A query cursor pointing to the
+                next batch of results. If there are no more results, this might
+                be None.
+            more: bool. If True, there are (probably) more results after this
+                batch. If False, there are no further results after this batch.
+    """
+    # The fetched skills will be filtered afterwards and filtering may result
+    # in having less number of skills than requested. Hence, fetching twice
+    # the number of requested skills will help reduce the number of datastore
+    # calls.
+    skill_summaries_models, new_urlsafe_start_cursor, more = (
+        skill_models.SkillSummaryModel.fetch_page(
+            2 * num_skills_to_fetch, urlsafe_start_cursor, sort_by))
+
+    skill_summaries = [
+        get_skill_summary_from_model(summary)
+        for summary in skill_summaries_models]
+    return skill_summaries, new_urlsafe_start_cursor, more
+
+
+def get_filtered_skill_summaries(
+        num_skills_to_fetch, status, classroom_name, keywords,
+        sort_by, urlsafe_start_cursor):
+    """Returns all the skill summary dicts after filtering.
+
+    Args:
+        num_skills_to_fetch: int. Number of skills to fetch.
+        status: str. The status of the skill.
+        classroom_name: str. The classroom_name of the topic to which the skill
+            is assigned to.
+        keywords: list(str). The keywords to look for
+            in the skill description.
+        sort_by: str. A string indicating how to sort the result.
+        urlsafe_start_cursor: str or None. The cursor to the next page.
+
+    Returns:
+        3-tuple(augmented_skill_summaries, new_urlsafe_start_cursor, more).
+        where:
+            augmented_skill_summaries: list(AugmentedSkillSummary). The list of
+                augmented skill summaries. The number of returned skills might
+                include more than the requested number. Hence, the cursor
```

Comment:
Is this fine though? Shouldn't you be returning a length equal to `num_skills_to_fetch` if you're passing that here anyway?

Reply:
Since some filtering is taking place, it's very much possible that we'll run short of the skills that were initially requested. So Sean and I decided it's better to reduce the DB calls and fetch maybe 2 or 3 times the requested number, Also the extra fetched skills are still being returned and saved in the frontend so no bandwidth or DB calls are being wasted.

Reply:
Ah ok, thanks for the explanation.